pkg/backend/malloc.go

```diff
@@ -0,0 +1,280 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (c) 2022-2024 Dell Inc, or its subsidiaries.
+// Copyright (C) 2023 Intel Corporation
+
+// Package backend implememnts the BackEnd APIs (network facing) of the storage Server
+package backend
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"path"
+	"sort"
+
+	"github.com/opiproject/gospdk/spdk"
+	pb "github.com/opiproject/opi-api/storage/v1alpha1/gen/go"
+	"github.com/opiproject/opi-spdk-bridge/pkg/utils"
+
+	"github.com/google/uuid"
+	"go.einride.tech/aip/fieldbehavior"
+	"go.einride.tech/aip/fieldmask"
+	"go.einride.tech/aip/resourceid"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+func sortMallocVolumes(volumes []*pb.MallocVolume) {
+	sort.Slice(volumes, func(i int, j int) bool {
+		return volumes[i].Name < volumes[j].Name
+	})
+}
+
+// CreateMallocVolume creates a Malloc volume instance
+func (s *Server) CreateMallocVolume(ctx context.Context, in *pb.CreateMallocVolumeRequest) (*pb.MallocVolume, error) {
+	// check input correctness
+	if err := s.validateCreateMallocVolumeRequest(in); err != nil {
+		return nil, err
+	}
+	// see https://google.aip.dev/133#user-specified-ids
+	resourceID := resourceid.NewSystemGenerated()
+	if in.MallocVolumeId != "" {
+		log.Printf("client provided the ID of a resource %v, ignoring the name field %v", in.MallocVolumeId, in.MallocVolume.Name)
+		resourceID = in.MallocVolumeId
+	}
+	in.MallocVolume.Name = utils.ResourceIDToVolumeName(resourceID)
+	// idempotent API when called with same key, should return same object
+	volume, ok := s.Volumes.MallocVolumes[in.MallocVolume.Name]
+	if ok {
+		log.Printf("Already existing MallocVolume with id %v", in.MallocVolume.Name)
+		return volume, nil
+	}
+	// not found, so create a new one
+	params := spdk.BdevMallocCreateParams{
+		Name:         resourceID,
+		BlockSize:    int(in.GetMallocVolume().GetBlockSize()),
+		NumBlocks:    int(in.GetMallocVolume().GetBlocksCount()),
+		MdSize:       int(in.GetMallocVolume().GetMetadataSize()),
+		MdInterleave: true,
```

Comment:
by default in SPDK it is false. Why do we need to have tn as `true` here?

Reply:
I need it to be `interleaved`.
If you think it's wrong in that way, I will add it as another field.

Reply:
I need to read what interleave means…

Reply:
Where the bdev keeps the MD. After each block (interleave) or on a dedicated region.